### PR TITLE
Add someLimit and everyLimit

### PR DESCRIPTION
**Pull request by megawac**
Merges the implementations of `some` and `every` and adds their respective limit functions (with partial short circuiting)

PS I think short circuiting should be done like lodash does it (allowing `each` to break):
  if `cb` resolves with `cb(false)` you break the loop (in lodash if you return false it breaks the loop)

Resolves #548

### CODE REVIEW

**Comment by megawac** [comment]
```sh
gyeates at boompad in ~/code/async on arr-testers [!$]
$ ./perf/benchmark.js -g some
Comparing 1.2.1 with current on Node v0.12.0
--------------------------------------
some - no short circuit- false(500) 1.2.1 x 1,838 ops/sec ±5.41% (23 runs sampled), 0.544ms per run
some - no short circuit- false(500) current x 1,899 ops/sec ±4.72% (25 runs sampled), 0.527ms per run
Tie
--------------------------------------
some - short circuit - true(500) 1.2.1 x 1,598 ops/sec ±10.93% (25 runs sampled), 0.626ms per run
some - short circuit - true(500) current x 2,007 ops/sec ±5.35% (26 runs sampled), 0.498ms per run
current is faster
--------------------------------------
current faster overall (1.02ms total vs. 1.17ms total)
current won more benchmarks (1 vs. 0)

gyeates at boompad in ~/code/async on arr-testers [!$]
$ ./perf/benchmark.js -g every
Comparing 1.2.1 with current on Node v0.12.0
--------------------------------------
every - no short circuit- true(500) 1.2.1 x 1,498 ops/sec ±13.85% (24 runs sampled), 0.668ms per run
every - no short circuit- true(500) current x 1,638 ops/sec ±5.26% (26 runs sampled), 0.611ms per run
Tie
--------------------------------------
every - short circuit - false(500) 1.2.1 x 1,589 ops/sec ±4.02% (25 runs sampled), 0.629ms per run
every - short circuit - false(500) current x 1,656 ops/sec ±16.95% (22 runs sampled), 0.604ms per run
1.2.1 is faster
--------------------------------------
current faster overall (1.21ms total vs. 1.3ms total)
1.2.1 won more benchmarks (1 vs. 0)
```

**Comment by aearly** [comment]
I've thought a lot about the best way to short circuit in the more general case.  `cb(false)` seems to generic for me -- most libraries just check for truthiness of the `err` value, so `false` might be a non-error, non-short-circuiting value for some people.

I've thought about defining a constant like `async.BREAK = {}` as a sort of a stop sentinel (or a poor man's Symbol).  You would pass this as the error in a `async.eachOf` iterator callback, and then the final callback for `eachOf` would handle it accordingly:

```js
function done(err) {
  if (err) {
    return callback(err === async.BREAK ? null : err);
  }
  //..
}
```

Seems a bit hackish, but I haven't thought of something better.  This type of strategy would also help in things like  `auto` or `waterfall` where you want to exit early, but you still need to cancel the remaining steps in the sequence to prevent memory leaks.

**Comment by aearly** [comment]
But :+1: for this otherwise, just waiting for CI.

File: lib/async.js [hunk 1]
```diff
@@ -449,35 +455,40 @@
     async.detect = doParallel(_detect);
     async.detectSeries = doSeries(_detect);
 
+    function _createTester(eachfn, check, defaultValue) {
+        return function(arr, limit, iterator, cb) {
+            function done() {
+                if (cb) cb(defaultValue);
```
Comment (aearly) [inline lib/async.js]: Should we add `cb = cb || noop` to prevent having to check for this every time?

File: package.json [hunk 2]
```diff
@@ -3,7 +3,7 @@
   "description": "Higher-order functions and common patterns for asynchronous code",
   "main": "lib/async.js",
   "author": "Caolan McMahon",
-  "version": "1.3.0",
+  "version": "1.2.1",
```
Comment (aearly) [inline package.json]: What's going on here?

**Comment by megawac** [comment]
> I've thought about defining a constant like async.BREAK = {} as a sort of a stop sentinel (or a poor man's Symbol).

Yeah, thats what underscore used to do and makes sense if false doesn't work

File: package.json (hunk 2 above)
Comment (megawac) [inline package.json]: whoops, didn't mean to commit that (it was just for benchmarking until #763 is figured out)

File: lib/async.js (hunk 1 above)
Comment (megawac) [inline lib/async.js]: This is only done at the last step, `noop` actually makes it more complicated because then at the other step (iterator) it has to check `if (cb === noop)`

**Comment by aearly** [comment]
> Yeah, thats what underscore used to do

Do you know why underscore moved away from it?

**Comment by megawac** [comment]
@aearly should I add a unit test to ensure some and every short circuit?

**Comment by aearly** [comment]
Yes -- always best to have tests for all the corner cases wherever possible.

**Comment by megawac** [comment]
> Do you know why underscore moved away from it?

Yeah, for performance reasons we implemented it as a `for` loop https://github.com/jashkenas/underscore/pull/1710

**Comment by megawac** [comment]
Should I add some unit tests to ensure they short circuit?

**Comment by aearly** [comment]
Ahh, I see --- it wasn't like lodash where it has a generic way to break, it was a special behavior for `_.some` and `_.every`

**Comment by aearly** [comment]
Why don't we save the short circuiting for 2.0 -- you're right -- it will have minor breaking changes.

**Comment by megawac** [comment]
IDK, I think the current behaviour can also be classified as a bug as people usually expect `some` and `every` to exit immediately (the only breaking change is the `test` function is called less)

**Comment by megawac** [comment]
This can also be used to implement `detect`, however I'll do that in a seperate PR

**Comment by aearly** [comment]
Actually, short circuiting isn't breaking at all.  Since we're using the parallel `_.eachOf` to implement `every`, the iterator will be called for every element regardless.  Short circuiting only applies to these new `everyLimit` variations.

**Comment by megawac** [comment]
Nope, `some` and `every` also short circuit because of this line https://github.com/caolan/async/pull/828/files#diff-688ede0ee2da1fa3d89f2fc485f39273R461

**Comment by megawac** [comment]
Wrong line https://github.com/caolan/async/pull/828/files#diff-688ede0ee2da1fa3d89f2fc485f39273R464

**Comment by aearly** [comment]
They exit early, but the iterators are still called for every element in the array.  The callbacks are just ignored once the relevant `true` or `false` is encountered.

**Comment by aearly** [comment]
Unless we're talking about a different subtle breaking change.

**Comment by megawac** [comment]
For synchronous iterators, `iterator` will only ever be called once as `eachOf` doesn't clal ensure sync AFAIK

**Comment by aearly** [comment]
Well, [this test](https://github.com/caolan/async/blob/master/test/test-async.js#L2222) didn't break, so I think we're okay.

**Comment by megawac** [comment]
It'll break if you remove the setTimeout :)

File: README.md [hunk 3]
```diff
@@ -172,8 +172,11 @@ Some functions are also available in the following forms:
 * [`reduce`](#reduce), [`reduceRight`](#reduceRight)
 * [`detect`](#detect), `detectSeries`
 * [`sortBy`](#sortBy)
-* [`some`](#some), [`every`](#every)
-* [`concat`](#concat), `concatSeries`
+* [`some`](#some)
+* [`someLimit`](#someLimit)
+* [`every`](#every)
+* [`concat`](#concat)
+* [`concatSeries`](#concatSeries)
```
Comment (charlierudolph) [inline README.md]: In #809, I updated how the docs are formatted where the series and limit functions no longer need their own documentation. Instead the series/limit is documented as whole and then the related commands are just mentioned with the documentation of the normal command.
 
These lines should become
```
* [`some`](#some), `someLimit`
* [`every`](#every), `everyLimit`
* [`concat`](#concat), `concatSeries`
```
And the documentation for someLimit should be removed and the following should be added to the `some` documentation.
```
__Related__

* someLimit(arr, limit, iterator, [callback])
```
Same for every.

Comment (megawac) [inline README.md]: Hi @charlierudolph, if you have some time would you mind submitting a pull request with your suggested documentation changes?
